### ISSUE TITLE
msgpackffi doesn't support sparsed arrays

### ISSUE THREAD

**Issue by pgulutzan** [tarantool/tarantool#1123]
I have Ubuntu 12.04 and Tarantool 1.6.7, pulled from source today.

I execute the following requests:

```
box.cfg{}
box.schema.create_space('T',{engine='memtx'})
box.space.T:create_index('T',{parts={1,'str',2,'str',3,'str'}})
box.space.T:insert{'1','2','3'}
box.space.T:insert{'1','2','0'}
box.space.T.index.T:select({'1','2'},{iterator='GT'})     -- select#1
box.space.T.index.T:select({'1','2'},{iterator='EQ'})     -- select#2
box.space.T.index.T:select({'1',nil,'5'},{iterator='EQ'}) -- select#3
box.space.T.index.T:select({'1',nil,'0'},{iterator='GT'}) -- select#4
```

Results:
- select#1 returns 0 rows
- select#2 returns 2 rows
- select#3 returns 2 rows
- select#4 returns 0 rows

**Assumption:** `{1,2}` should be equivalent to `{1,2,nil}`, that is,
if trailing fields are not specified then they are assumed to be nil.

**Assumption:** We do not expect that index-field values will ever include nil.

Alternatives for how to treat nil when comparing to an index-field value:
1. `nil` is "equal to" anything, therefore if `iterator = EQ|LE|GE|REQ|ALL`
   then the comparison must be true, but if `iterator = GT|LT`
   then the comparison must be false
2. `nil` is "greater than" anything, therefore if `iterator = LT|LE|ALL`
   then the comparison must be true, but if `iterator = EQ|GT|GE|REQ`
   then the comparison must be false
3. `nil` "matches" anything, therefore for all iterators
   the comparison must be true
4. `nil` "does not match" anything, therefore for all iterators
   the comparison must be false

Notice my terminology: when I say "match" I do not mean "equal to".

**Claims:**
Regardless what alternative you pick, the four select results
are not following it consistently, and select#3 is always wrong.

Incidentally the manual currently says

> If the number of passed field-values is less than
>  the number of fields in the primary key, then only the passed field-values are compared, so select{1,2} will match a tuple whose primary key is {1,2,3}.

but that is not what is happening, and is not clear.

This is a report about memtx storage engine only.

**Comment by rtsisyk** [comment]
Hi Peter,

Unfortunately, we haven't implemented support for  `nil` in indexes yet. Thanks for the such detailed explanation of `nil` behavior, though. I hope this feature will be implemented  in 1.7.x (see #1257)

Currently Tarantool should raise an error about invalid data type (Supplied key type of part 1 does not match index part type: expected STR). 

```
tarantool> box.space.T.index.T:select({'1',nil,'5'},{iterator='EQ'})
---
- - ['1', '2', '0']
  - ['1', '2', '3']
...

tarantool> box.space.T.index.T:select({'1',require('msgpack').NULL,'5'},{iterator='EQ'})                                                                                                     
---
- error: 'Supplied key type of part 1 does not match index part type: expected STR'
...
```

It seems that your problem is caused by a bug in msgpackffi code.
I'll fix it.

See also #1167.

**Comment by rtsisyk** [comment]
I need a decision on #1167 to implement the same logic in msgpackffi.

Blocked by #1167

**Comment by rtsisyk** [comment]
@pgulutzan  please use `msgpack.NULL` instead of `nil`:

```
NULL = require('msgpack').NULL
box.cfg{}
box.schema.create_space('T',{engine='memtx'})
box.space.T:create_index('T',{parts={1,'str',2,'str',3,'str'}})
box.space.T:insert{'1','2','3'}
box.space.T:insert{'1','2','0'}
```

```
tarantool> box.space.T.index.T:select({'1', nil,'5'},{iterator='EQ'}) -- select#3
---
- - ['1', '2', '0']
  - ['1', '2', '3']
...

tarantool> box.space.T.index.T:select({'1', nil,'0'},{iterator='GT'}) -- select#4
---
- []
...
```

`nil` has been lost ....


```
tarantool> box.space.T.index.T:select({'1', NULL,'5'},{iterator='EQ'}) -- select#3
---
- error: 'Supplied key type of part 1 does not match index part type: expected STR'
...

tarantool> box.space.T.index.T:select({'1', NULL,'0'},{iterator='GT'}) -- select#4
---
- error: 'Supplied key type of part 1 does not match index part type: expected STR'
...
```


I think we should add more information about `nil` and `NULL` in Lua.

**Comment by rtsisyk** [comment]
One more example to illustrate this bug:

```
tarantool> msgpackffi = require('msgpackffi')
---
...
tarantool> msgpackffi.decode(msgpackffi.encode({1, nil, 3}))
---
- [1] <!-- nil has been lost ....
- 3
...

tarantool> msgpackffi.decode(msgpackffi.encode({1, msgpackffi.NULL, 3}))
---
- [1, null, 3]
- 5
...
```

**Comment by rtsisyk** [comment]
I checked msgpackffi code again. Unfortunately, it is not possible to implement luaL_serializer logic in msgpackffi, because `pairs()` and `next()` functions break JIT trace. The core idea of msgpackffi was to allow JIT in selects. Since msgpackffi cannot be fixed, I close this ticket. Sorry.

Documentation => #1300

**Comment by rtsisyk** [comment]
I discussed with @kostja . msgpackffi is buggy and we have no idea how to fix it.
This ticket should be open for a while, until this problem will be fixed.

**Comment by rtsisyk** [comment]
msgpackffi should support `__serialize`. A real test case from a customer:

```
tarantool> t = box.tuple.new({1, 2, 3})
---
...

tarantool> map = setmetatable({}, { __serialize = 'map' })
---
...

tarantool> map
---
- {}
...

tarantool> t:update({{'=', 3, map}}) -- msgpack FFI
---
- [1, 2, []] <!-- an array has been inserted
...

tarantool> box.space.memtx:update(1, {{'=', 3, map}}) <!-- msgpack Lua/C
---
- [1, 2, {}] <!-- a map has been inserted
...
```

**Comment by rtsisyk** [comment]
Fixed as part of #2036 using `pairs()`.  `pairs()` is [NYI](http://wiki.luajit.org/NYI), but it doesn't matter anymore. LuaJIT has never had a working JIT.
